linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -0,0 +1,2206 @@
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.isUndef)
+      },
+      List(KindBoxedBoolean) -> { () =>
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.typeTest(BooleanRef))
+      },
+      List(KindBoxedCharacter) -> { () =>
+        fb += LocalGet(valueParam)
+        val structTypeID = genTypeID.forClass(SpecialNames.CharBoxClass)
+        fb += RefTest(RefType(structTypeID))
+      },
+      List(KindBoxedByte) -> { () =>
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.typeTest(ByteRef))
+      },
+      List(KindBoxedShort) -> { () =>
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.typeTest(ShortRef))
+      },
+      List(KindBoxedInteger) -> { () =>
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.typeTest(IntRef))
+      },
+      List(KindBoxedLong) -> { () =>
+        fb += LocalGet(valueParam)
+        val structTypeID = genTypeID.forClass(SpecialNames.LongBoxClass)
+        fb += RefTest(RefType(structTypeID))
+      },
+      List(KindBoxedFloat) -> { () =>
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.typeTest(FloatRef))
+      },
+      List(KindBoxedDouble) -> { () =>
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.typeTest(DoubleRef))
+      },
+      List(KindBoxedString) -> { () =>
+        fb += LocalGet(valueParam)
+        fb += Call(genFunctionID.isString)
+      },
+      // case KindJSType => call typeData.isJSClassInstance(value) or throw if it is null
+      List(KindJSType) -> { () =>
+        fb.block(RefType.anyref) { isJSClassInstanceIsNull =>
+          // Load value as the argument to the function
+          fb += LocalGet(valueParam)
+
+          // Load the function reference; break if null
+          fb += LocalGet(typeDataParam)
+          fb += StructGet(genTypeID.typeData, isJSClassInstance)
+          fb += BrOnNull(isJSClassInstanceIsNull)
+
+          // Call the function
+          fb += CallRef(genTypeID.isJSClassInstanceFuncType)
+          fb += Call(genFunctionID.box(BooleanRef))
+          fb += Return
+        }
+        fb += Drop // drop `value` which was left on the stack
+
+        // throw new TypeError("...")
+        fb ++= ctx.getConstantStringInstr("TypeError")
+        fb += Call(genFunctionID.jsGlobalRefGet)
+        fb += Call(genFunctionID.jsNewArray)
+        fb ++= ctx.getConstantStringInstr(
+          "Cannot call isInstance() on a Class representing a JS trait/object"
+        )
+        fb += Call(genFunctionID.jsArrayPush)
+        fb += Call(genFunctionID.jsNew)
+        fb += ExternConvertAny
+        fb += Throw(genTagID.exception)
+      }
+    ) { () =>
+      // case _ =>
+
+      // valueNonNull := as_non_null value; return false if null
+      fb.block(RefType.any) { nonNullLabel =>
+        fb += LocalGet(valueParam)
+        fb += BrOnNonNull(nonNullLabel)
+        fb += GlobalGet(genGlobalID.bFalse)
```

Comment:
Argh, this confused me, because for normal paths, we box at the end. Not clear if there is something we can do about it though.

Probably OK to leave it as is.